doc/media-settings/Media-based-Port-settings.md

```diff
@@ -0,0 +1,313 @@
+                        "0x4321",
+
+                        "0x4321"
+
+                    ],
+
+                    "idriver": [
+
+                        "0x2",
+
+                        "0x3"
+
+                    ],
+
+                    "ipredriver": [
+
+                         "0x3",
+
+                         "0x1"
+
+                    ]
+
+                },
+
+                "DELL-00-11-22": {
+
+                    "preemphasis": [
+
+                        "0x1311",
+
+                        "0x321c",
+
+                    ],
+
+                    "idriver": [
+
+                        "0x1",
+
+                        "0x2",
+
+                    ],
+
+                    "ipredriver": [
+
+                        "0x2",
+
+                        "0x1",
+
+                    ]
+
+                }
+
+            },
+
+            "40GBASE-CR4-3M": {
+
+                "Default": {
+
+                    "preemphasis": [
+
+                        "0x1021",
+
+                        "0x3021",
+
+                    ],
+
+                    "idriver": [
+
+                        "0x4",
+
+                        "0x4"
+
+                    ],
+
+                    "ipredriver": [
+
+                        "0x4",
+
+                        "0x4"
+
+                    ]
+
+                },
+
+                "DELL": {
+
+                    "preemphasis": [
+
+                        "0x1311",
+
+                        "0x1312
+
+                    ],
+
+                    "idriver": [
+
+                        "0x2",
+
+                        "0x2"
+
+                    ],
+
+                    "ipredriver": [
+
+                        "0x2",
+
+                        "0x2"
+
+                    ]
+
+                }
+
+            }
+
+        }
+
+    }
+
+}
+
+![](https://github.com/dgsudharsan/SONiC/blob/hld_media/doc/media-settings/event_flow.png)
+
+Flow:
+=====
+
+              When a media is detected in xcvrd daemon, it constructs the media key identifier string as discussed above and searches the media_settings.json file. On finding a matching media key entry the vendor key is constructed and if found the key value pairs are fetched. The xcvrd daemon notifies these key value pairs along with the alias port to portsorch. The portsorch task will convert the alias to SAI object Id and notifies the syncd by doing port attribute set of corresponding settings. Syncd invokes the SAI port set those attribute with specified values. The SAI implementation then programs the attribute into the hardware.
+
+              The notification of hardware profile from xcvrd to portsorch task will be done during initialization and during media detect event. This is not required during media removal event. Since the media settings are required only for the proper functioning of the optics or DAC, the handling can be restricted to media insert event alone and no action needs to be taken during media removal.
```

Comment:
during switch initialization, xcvrd will wait for all the ports configured then startup itself, so maybe need to find another way for the event detection in this phase.

Reply:
We plan to have a one-shot task during bootup which will notify porsorch, the media settings for all ports based on media detected. Once xcvrd starts, it will start notifying the media settings during transceiver change events.